private-identity-audit/src/lib.rs

```diff
@@ -0,0 +1,254 @@
+    ///
+    /// # Arguments
+    /// * `initial_message`: Initial messages of ZKP proofs of CDD_ID.
+    /// * `final_response`: Final responses of ZKP proofs of CDD_ID.
+    /// * `challenge`: The ZKP challenge generated by PUIS.
+    /// * `cdd_id`: The CDD_ID read from the chain.
+    /// * `committed_cdd_id`: The committed CDD_ID received from CDD Provider.
+    /// * `committed_cdd_id_second_half`: The committed version of the second half of CDD_ID
+    ///    received from CDD Provider.
+    /// * `verifier_secrets`: The PUIS secrets generated in the second step of the protocol.
+    /// * `rng`: Cryptographically secure random number generator.
+    fn verify_proofs(
+        initial_message: Proofs,
+        final_response: ProverFinalResponse,
+        challenge: Scalar,
+        cdd_id: RistrettoPoint,
+        committed_cdd_id: RistrettoPoint,
+        committed_cdd_id_second_half: RistrettoPoint,
+        verifier_secrets: VerifierSecrets,
+        re_encrypted_uids: EncryptedUIDs,
+    ) -> Fallible<()>;
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        prover::{generate_bliding_factor, FinalProver, InitialProver},
+        uuid_to_scalar,
+        verifier::{gen_random_uuids, Verifier, VerifierSetGenerator},
+        ChallengeGenerator, ChallengeResponder, InvestorID, ProofGenerator, ProofVerifier,
+    };
+    use confidential_identity::pedersen_commitments::PedersenGenerators;
+    use cryptography_core::curve25519_dalek::scalar::Scalar;
+    use rand::{rngs::StdRng, SeedableRng};
+
+    #[test]
+    fn test_success_end_to_end() {
+        let mut rng = StdRng::from_seed([10u8; 32]);
+        let pg = PedersenGenerators::default();
+
+        // Private input of the Verifier.
+        let private_uid_set: Vec<Scalar> = gen_random_uuids(100, &mut rng)
+            .into_iter()
+            .map(|uuid| uuid_to_scalar(uuid))
+            .collect();
+
+        // Verifier shares one of its uids with the Prover.
+        let investor = InvestorID {
+            uid: private_uid_set[0],
+            did: Scalar::random(&mut rng),
+        };
+
+        // Prover generates cdd_id and places it on the chain.
+        let cdd_id = pg.commit(&[
+            investor.uid,
+            investor.did,
+            generate_bliding_factor(investor.uid, investor.did),
+        ]);
+
+        // P -> V: Prover generates and sends the initial message.
+        let (prover_secrets, proofs, committed_cdd_id, committed_cdd_id_second_half) =
+            InitialProver::generate_initial_proofs(investor, &mut rng).unwrap();
+
+        // V -> P: Prover sends `proofs` and Verifier returns a list of 10 uids and the challenge.
+        let (verifier_secrets, encrypted_uids, challenge) = VerifierSetGenerator
+            .generate_encrypted_set_and_challenge(private_uid_set, Some(100), &mut rng)
+            .unwrap();
+
+        // P -> V: Verifier sends the encrypted_uids and the challenge to the Prover.
```

Comment:
encrypted uid? that's a confusing choice of word.

Reply:
Yeah, I used its proper name `committed uids` and `recommitted uids` inside the function. I made this `encrypted` because I thought the users of the library would be more comfortable with it and it is technically an encryption. But I can change it if it is confusing.